Microsoft.Toolkit.HighPerformance/Buffers/StringPool.cs

```diff
@@ -0,0 +1,199 @@
+            }
+
+            if (entriesPerBucket <= 0)
+            {
+                ThrowArgumentOutOfRangeException(nameof(entriesPerBucket));
+            }
+
+            Span<Bucket> span = this.buckets = new Bucket[numberOfBuckets];
+
+            // We preallocate the buckets in advance, since each bucket only contains the
+            // array field, which is not preinitialized, so the allocations are minimal.
+            // This lets us lock on each individual buckets when retrieving a string instance.
+            foreach (ref Bucket bucket in span)
+            {
+                bucket = new Bucket();
+            }
+
+            NumberOfBuckets = numberOfBuckets;
+            EntriesPerBucket = entriesPerBucket;
+        }
+
+        /// <summary>
+        /// Gets the default <see cref="StringPool"/> instance.
+        /// </summary>
+        public static StringPool Default { get; } = new StringPool();
+
+        /// <summary>
+        /// Gets the total number of buckets in use.
+        /// </summary>
+        public int NumberOfBuckets { get; }
+
+        /// <summary>
+        /// Gets the maximum number of <see cref="string"/> entries stored in each bucket.
+        /// </summary>
+        public int EntriesPerBucket { get; }
+
+        /// <summary>
+        /// Gets a cached <see cref="string"/> instance matching the input content, or creates a new one.
+        /// </summary>
+        /// <param name="span">The input <see cref="ReadOnlySpan{T}"/> with the contents to use.</param>
+        /// <returns>A <see cref="string"/> instance with the contents of <paramref name="span"/>, cached if possible.</returns>
+        public string GetOrAdd(ReadOnlySpan<char> span)
+        {
+            if (span.IsEmpty)
+            {
+                return string.Empty;
+            }
+
+            int bucketIndex = span.Length % NumberOfBuckets;
+
+            Bucket bucket = this.buckets.DangerousGetReferenceAt(bucketIndex);
+
+            lock (bucket)
+            {
+                return bucket.GetOrAdd(span, EntriesPerBucket);
+            }
+        }
+
+        /// <summary>
+        /// Resets the current instance and its associated buckets.
+        /// </summary>
+        public void Reset()
+        {
+            foreach (Bucket bucket in this.buckets)
+            {
+                lock (bucket)
+                {
+                    bucket.Reset();
+                }
+            }
+        }
+
+        /// <summary>
+        /// A configurable bucket containing a group of cached <see cref="string"/> instances.
+        /// </summary>
+        private sealed class Bucket
+        {
+            /// <summary>
+            /// A bitmask used to avoid branches when computing the absolute value of an <see cref="int"/>.
+            /// This will ignore overflows, as we just need this to map hashcodes into the valid bucket range.
+            /// </summary>
+            private const int SignMask = ~(1 << 31);
```

Comment:
FWIW, this is same as `int.MinValue`. Up to you whether that is clearer or not, tho

Reply:
In my mind this is actually clearer as it makes it explicit which bits you're actually interested in (first is 0, all others set to 1), whereas with just `int.MinValue` it might not be immediately clear what the representation is, considering that also takes into account how 2-complement numbers are represented. Imho I'd leave this like this 🤔